### DISCUSSION TITLE
Remarkable software-FYI MAX98357A I2S amplifier also works for DAC

### DISCUSSION THREAD

**Discussion by bmillier** [probonopd/MiniDexed#179] [General]
MiniDexed is remarkable software. I got it up and running in no time using an Adafruit Raspi speaker Hat (lPimoroni Speaker pHat also uses the same two MAX98357 I2S amplifiers). I was interested in running it from my DAW PC app and just added a small circuit using the 6N138 opto-coupler feeding the Pi's RX line. This works perfectly.
  If I could could suggest something- How to add the folder for the voice banks, as well as the naming convention needed, can now be found here in the discussion threads, but it wouldn't hurt to add that info to the wiki- maybe in the Menu section.
   Once I saw how to do it, it was easy to import syx files from the yamahablackbox site that was cited. However, I found that none of the TX802 sysex files that I had (from way back), would work. Turns out they are a bit larger than the DX7 banks on the yamahablackbox site and actually include 2 sysex file structures, I don't know what the first one is,but it ends in F7h like all sysex files. Following that is the bank voice file starting with F0h and ending in F7h. When I edited out the first sysex structure, MiniDexed read the TX802 voice bank perfectly.
  I'm considering designing a PCB Raspi shield containing the DIN MIDI input circuit, a header for the LCD and contrast pot, and header for the Rotary encoder. Adafruit has a nice UDA1334 breakout board for $7 US , which I would mount for the DAC  function (most DACs are TSSOP which I find tricky to solder in). I've used this breakout board with Teensy 4.0's standard I2S driver, and it works fine, so it should also work with Raspi.

**Comment by probonopd** [comment]
Hello @bmillier, thanks for letting us know.

> MAX98357

Interesting! How is the sound quality?

According to https://cdn-learn.adafruit.com/assets/assets/000/037/882/original/raspberry_pi_schem.png, 2 of those Mono chips can be used for Stereo. Does that work for you with MiniDexed?

> I was interested in running it from my DAW PC app and just added a small circuit using the 6N138 opto-coupler feeding the Pi's RX line. This works perfectly.

Can you please post the *exact* working circuit in https://github.com/probonopd/MiniDexed/discussions/170. That would be tremendous! We have been trying to get this to work but possibly did it wrong so far.

> How to add the folder for the voice banks

The README.md mentions `getsysex.sh` but apparently we need to be more verbose than that. Have added a new page to the Wiki:
https://github.com/probonopd/MiniDexed/wiki/Files

Thanks for your information on how to use TX802 `.syx` files. I have opened a feature request for this, #180. Maybe someone would like to implement it.

> I'm considering designing a PCB Raspi shield containing the DIN MIDI input circuit, a header for the LCD and contrast pot, and header for the Rotary encoder.

Please join the discussion at https://github.com/probonopd/MiniDexed/issues/88. We are collecting the requirements there. Your DIN MIDI input circuit would be cool. Also a DIN MIDI _output_ circuit!

> Adafruit has a nice UDA1334 breakout board for $7 US , which I would mount for the DAC function (most DACs are TSSOP which I find tricky to solder in). I've used this breakout board with Teensy 4.0's standard I2S driver, and it works fine, so it should also work with Raspi.

Theoretically the UDA1334 should work, but I don't think anyone has tested it yet.

We should also have footprints for the [GY PCM5102 DAC module](https://github.com/dwhinham/mt32-pi/wiki/GY-PCM5102-DAC-module) and the [CJMCU 5102 DAC module](https://github.com/dwhinham/mt32-pi/wiki/CJMCU-5102-DAC-module) which can be had very inexpensively.

**Comment by bmillier** [comment]
Hi probonopd:   The Adafruit board has 2 MAX98357s on it configured for Left, Right. So far, I only hooked up 1 small Polk Audio speaker enclosure, so I am only hearing 1 channel, but I've used this board before and it puts out stereo. The Polk Audio speakers aren't great (brand notwithstanding) as they are just surplus Rear speakers from a TV home theatre setup. But, the quality is pretty good- I don't expect it to sound as good as what I get from my music recording setup with a quality amp and large Cerwin-Vega speakers!
. Using the MAX98357's would really only make sense if you wanted a free-standing keyboard with built in amp/speakers. The Adafruit board was all I had on hand to try out immediately and I thought others might be in the same position.
 I will post the DIN MIDI IN diagram tomorrow- after I figure out how to incorporate it into the comment field. I think I saw in a thread someone tried a 6N137. I believe they are too slow- I've always used the 6N138 and it always works flawlessly. However, you do need to feed its Vcc pin with 5V- not 3.3V. And, the pull-up resistor (1K) on the output Collector has to go to 3.3V (NOT 5V)

Yes, I saw getsysex.sh. That showed me the URL for the DX7 voice libraries and I noticed the sysex/voice folder needed. But, I didn't pay attention to the 00000x_  prefix, and you need that. Also, it didn't help that my TX802 sysex files were incompatible (as-is) Those were the first files I tried. My TX802 sysex files had primary filenames >8 chars, and I think that was an issue too.
  You mention a DIN MIDI Output.  Does the MiniDexed send out any MIDI messages? Maybe you mean a MIDI THRU, which just allows one to daisy-chain more than one MIDI instrument on a single MIDI port. MIDI THRU is easy to implement- just need one logic chip (or maybe a transistor) for the driver. MIDI OUT also uses the same circuit, but connected to the  Raspi Tx pin.
   The two DAC modules you cite would be great as they use the PCM5102. Since the pandemic/logistics problems came up, I haven't even tried to order anything directly from China (I'm in Canada). That's why I considered the Adafruit UDA1334 since I can get it from Mouser quickly with no shipping charge if I add it to a regular order. But I'll check these two breakouts further now.
I haven't yet hooked up the Raspi/MiniDexed to my logic analyser to see if the I2S stream is 16-bit or is greater. If its 16-bit, then the higher resolution of the PCM5102 (32-bit) wouldn't make too much difference. That said, if I could get one of these PCM5102 boards easily, I'd use it for sure.
 I'll look at thread #88 to see what's happening.
Thanks for your very prompt response. I just got through eating dinner and your 2 email notifications were already waiting for me.

**Reply by probonopd** [reply]
We are doing MIDI TRHU in software. For that we'd need the MIDI OUT circuit. And who knows, maybe some day we'll add a step sequencer or some other functionality that makes use of it.

**Reply by craigyjp** [reply]
> We are doing MIDI TRHU in software. For that we'd need the MIDI OUT circuit. And who knows, maybe some day we'll add a step sequencer or some other functionality that makes use of it.

I built a MIDI in/out, didn't add the through as I hope to get it done in software, I started looking at how to code that yesterday but my wife insisted I did the gardening instead. You can change the 74HCT245 for a 74HC14 but you would need to use two inverters in series for the signal to be correct.

![midi](https://user-images.githubusercontent.com/51026305/165526339-ae2911a5-ed11-4576-b18d-126a5031a63e.jpg)
.

**Reply by probonopd** [reply]
The MIDI THROUGH in software is _already_ coded. So no need for a MIDI THRU port imho. The MIDI OUT can act as such.

**Reply by bmillier** [reply]
@ craigyjp + ,probonopd   I am curious about why you feel you need software for the MIDI THRU. MIDI THRU is just a  MIDI signal replicated from the incoming MIDI IN (so you can daisy-chain multiple tone generators on the same MIDI line). This is what you are showing in the diagram above- the logic level MIDI IN is sent back out via the 74HCT245N. So, even without the Raspi present, the MIDI In would get sent through to the MIDI THRU. You DON'T want to replicate the MIDI IN signal to the MIDI OUT. This isn't done in commercial tone generators etc (unless you configure it to have the MIDI OUT also act as MIDI THRU).  The only signals that the MiniDexed should send to MIDI OUT would be a) if you implemented a sysex dump function to dump out the bank/voice data to a PC/DAW.  b) Some commercial instruments will respond to a sysex "device identifier" query with a sysex message which identifies what particular tone generator it is. That sysex response would also go out the MIDI OUT port. 
  If you send all MIDI IN signals to MIDI OUT, then you will get a "digital" feedback loop in some cases when you are hooked up to a DAW, and are using sysex messages for dumps, etc.
@probonopd - I said I'd submit a drawing of my DIN MIDI IN circuit. Craigyjp's circuit has that plus the Thru and Out, so no sense in me duplicating it. I happen to use the 6N138 opto as I use them for other purposes and they are fast enough for MIDI. But, craigyjps 4N35 is a common opto currently and its speed specs are as good, and it must be working for craig.

**Reply by craigyjp** [reply]
I didn't actually build the midi through part of the design as I assumed midi thru was switched in software from a menu, them I realized it wasn't. 

So back to your point, a switchable midi thru would be useful as it allows you to use the unit as a usb to midi din converter and vice versa.

If this isn't a requirement then my full circuit gives you a thru regardless. But only for the midi din. 

I would prefer to have all options available.

The original design had a 6n138, but I only have 4n35 and pc900 chips in my parts bin and I have used the 4n35 successfully in many projects.

**Reply by bmillier** [reply]
Yes- if you are using the USB MIDI, then the switchable MIDI Thru allows for a USB MIDI to DIN MIDI converter. I can't use the USB MIDI IN on the MiniDexed with my DAW, as I need a MIDI device not a MIDI host, as you know. So, I had forgotten about this possibility. 
  I could never find PC900s years ago when I was building MIDI circuits, but had used 6N138s a lot for other things. But, 6N138  need 5V for Vcc, so 1 extra line needed from the Raspi. Many other people have tried various 4Nxx and HA11xx optos but many are too slow or marginal.

**Reply by probonopd** [reply]
> I am curious about why you feel you need software for the MIDI THRU

In my case, I attach a USB MIDI keyboard to MiniDexed, and I want its signal to come out of the serial port of MiniDexed, which in my case I put into a Raspberry Pi Pico attached to a tablet. So this is why I need MIDI THRU in software anyway (no Mini-DIN MIDI used in my case at all.)

**Reply by craigyjp** [reply]
Everyone's use case is different, hence why it would be nice to have all the options.

**Comment by bmillier** [comment]
Hi probonopd:
   I found many sources for GY PCM5102- I can get it on Amazon with free shipping but it will take 2-3 weeks if everything goes well. I'll place an order- I'm curious to see if electronics parts direct from China is practical these  days.
Cheers

**Reply by probonopd** [reply]
One shop had cancelled my order, but I don't know whether everything is closed down right now.

**Comment by craigyjp** [comment]
Yeah, the only way to use the pi is with din midi when using a DAW. You
could build the midi thru part of the circuit I provided and still have the
software midi thru option, just never activate it.

If we get sysex and midi dumps working then full din midi in/out will be
needed to edit the patches etc.

**Comment by bmillier** [comment]
On the Teensy 4x MCUs that I use, you can set up its USB port as a MIDI DEVICE with both IN and OUT profiles. Or, you can set it up as a host. I guess it must be more difficult to do this with the Raspi. I have no bare-metal programming experience with the Raspi - yet

**Comment by craigyjp** [comment]
It's two separate usb ports on the Teensy, the main usb is in/out and the
extra usb port is the usb host port AFAIK

**Comment by bmillier** [comment]
Yes, it is two separate ports. The Teensy's main USB port acts as a MIDI DEVICE, both in and out. So, I can hook my Teensy-based MIDI devices directly to my DAW via USB. I haven't needed to use the Teensy's USB host port for anything yet.

**Comment by craigyjp** [comment]
I use them both extensively in my projects. Best thing on the T3.6 and T4.x

**Comment by bmillier** [comment]
Yes- The Teensy has great software support and a great forum. Plus, the 4x are so fast on floating point, CMSIS DSP math etc.
I use Visual Micro/Visual Studio as my IDE. What do you use?

**Reply by craigyjp** [reply]
I'm still using the Arduino IDE, I've never got in to using anything else, I tried some other environment who's name escapes me but got stuck with how to actually use it. So I stick with the Arduino IDE and it serves my purposes even if it is a bit clunky.

**Comment by bmillier** [comment]
Yes, maybe  PlatformIO. It came out a while back. It supports many MCUs including those in the Arduino family + Teensy. I tried it and didn't like it as much as VM/VS. Visual Micro had a bit of a steep learning curve, but was free at first, when I got it. But, once I got used to it, it was so powerful, due to all the features of Visual Studio, that I cant' do without it now. Arduino themselves have improved their IDE lately- I tried a beta version but it wasn't nearly as good as Visual Micro, so I didn't pursue it.

**Comment by craigyjp** [comment]
Yes, platform Io, I didn't get on with it at all
